### DISCUSSION TITLE
Having trouble making a formula work on both macOS and Linux

### DISCUSSION THREAD

**Discussion by mainrs** [Homebrew/discussions#1085] [Linux] [answered]
It seems that Linux binaries are meant to be statically linked. However, I couldn't find any instructions on how to run different install commands depending on what the OS is. I have zero Ruby experience, so maybe the solution is quite obvious. For completeness, I just need to pass an argument to `cargo` if I am building on linux.

Is a better approach to have two different taps, one per OS? It seems that that is what homebrew does (`linuxbrew-core` and `homebrew-core`). If that is the case, how does homebrew know which to use? 

```rb
class GhLabelsCli < Formula
  desc "CLI for managing labels in existing GitHub repositories"
  homepage "https://github.com/SirWindfield/gh-labels-cli"
  url "https://github.com/SirWindfield/gh-labels-cli/archive/v0.1.8.tar.gz"
  sha256 "ad66dedf7fcd42d9c3cac0430a6e3d667cf4f197103392aa09635e5e2f3a2f84"
  license any_of: ["Apache-2.0", "MIT"]
  head "https://github.com/SirWindfield/gh-labels-cli.git"

  depends_on "rust" => :build
  # depends_on :macos

  def install
    system "cargo", "install", *std_cargo_args
  end

  test do
    system "#{bin}/gh-labels", "help"
  end
end
```

**Comment by SMillerDev** [comment, marked as answer]
> It seems that Linux binaries are meant to be statically linked.

They are not, homebrew prefers shared binaries whenever possible.

> I couldn't find any instructions on how to run different install commands depending on what the OS is.

Look in homebrew-core for usage of `on_linux` and `on_macos`

**Reply by mainrs** [reply]
> They are not, homebrew prefers shared binaries whenever possible.

Then what does this error message mean? Because I interpreted it that way:

```log
==> brew linkage --test sirwindfield/tap/gh-labels-cli
==> FAILED
Unwanted system libraries:
  /lib/x86_64-linux-gnu/libz.so.1
```

> Look in homebrew-core for usage of `on_linux` and `on_macos`

Thanks!

**Reply by gromgit** [reply]
That simply says that your formula links with system libraries other than the `glibc` and `gcc` ones, so any bottles created from this formula can break on Linux boxes that don't have the same versions of said libraries installed...or don't have them at all.

The fix is simple:
```
depends_on  "zlib"
```

**Reply by carlocab** [reply]
This is probably better:
```ruby
uses_from_macos "zlib"
```

**Reply by mainrs** [reply]
> That simply says that your formula links with system libraries other than the `glibc` and `gcc` ones, so any bottles created from this formula can break on Linux boxes that don't have the same versions of said libraries installed...or don't have them at all.
> 
> The fix is simple:
> 
> ```
> depends_on  "zlib"
> ```

Ah! Thanks for letting me know!
